### ISSUE TITLE
为什么SFTDataset 中  loss_mask会这么设计呢？

### ISSUE THREAD

**Issue by lwdnxu** [jingyaogong/minimind#476]
大佬们好，有个问题想咨询一下，为什么SFTDataset 中  loss_mask只关注assistant部分，不关注system 和user呢?  这里和 loss_mask等于attention_mask 有什么区别呢？

```python
    def _generate_loss_mask(self, input_ids):
        loss_mask = [0] * len(input_ids)
        i = 0
        while i < len(input_ids):
            if input_ids[i:i + len(self.bos_id)] == self.bos_id:
                start = i + len(self.bos_id)
                end = start
                while end < len(input_ids):
                    if input_ids[end:end + len(self.eos_id)] == self.eos_id:
                        break
                    end += 1
                for j in range(start + 1, min(end + len(self.eos_id) + 1, self.max_length)):
                    loss_mask[j] = 1
                i = end + len(self.eos_id) if end < len(input_ids) else len(input_ids)
            else:
                i += 1
        return loss_mask


```

**Comment by zhenyu-02** [comment]
这是一个经典的 LLM 面试题，直观的理解是：
我们希望让模型学会“如何作为助手（assistant）进行回应”，而不是其他内容。
最早应该是 Openai 的 InstructGPT 提出的，原文描述：
"We begin by fine-tuning GPT-3 to follow instructions using supervised learning (SFT), where we collect demonstrations of desired model behavior and **train the model to imitate them**."
https://arxiv.org/abs/2203.02155

**Comment by zhenyu-02** [comment]
对于 attention mask，我的理解是避免Self Attention 对 padding token 计算注意力分数（分配一个极小值，在 softmax 后接近 0）。但好像在代码里没有被用到。
反向传播时的 loss mask 也是避免 padding 部分的损失污染参数。
两个 mask 都是不希望模型关注空白部分的注意力。

**Comment by gaochao0609** [comment]
在有监督微调（SFT）任务中，`SFTDataset` 的 `loss_mask` 只关注 `assistant` 部分而忽略 `system` 和 `user` 部分，以及它与 `attention_mask` 的区别，本质上是由两者的设计目标和作用机制决定的。


### 1. 为什么 `loss_mask` 只关注 `assistant` 部分？
`loss_mask` 的核心作用是**指定哪些位置的预测需要计入损失函数**，从而引导模型学习特定目标。在 SFT 任务中，我们的目标是让模型学习“根据用户输入生成合理回答”，因此：

- **`assistant` 部分是模型需要学习生成的内容**：用户的问题（`user`）和系统提示（`system`）是模型的“输入上下文”，而助手的回复（`assistant`）是模型需要学习输出的“目标结果”。只有 `assistant` 部分的预测错误需要被惩罚（计入损失），才能让模型学会生成符合预期的回答。
  
- **`user` 和 `system` 部分是输入而非目标**：如果对 `user` 或 `system` 部分计算损失，模型会被误导去“预测输入内容”（比如重复用户的问题），而不是专注于生成回答。这与 SFT 的目标（学习生成回答）相悖。

从代码实现来看，`_generate_loss_mask` 通过识别 `assistant` 的起始标记（`self.bos_id = '<|im_start|>assistant'`）和结束标记（`self.eos_id = '<|im_end|>'`），仅对这两个标记之间的内容设置 `loss_mask=1`（需要计算损失），其他部分（`user`/`system`）则为 `0`（忽略损失）。


### 2. `loss_mask` 与 `attention_mask` 的区别
两者的作用完全不同：

| 维度         | `loss_mask`                          | `attention_mask`                    |
|--------------|--------------------------------------|-------------------------------------|
| **作用目标** | 控制哪些位置的预测损失被计入总损失   | 控制模型注意力机制中哪些 token 有效 |
| **核心逻辑** | 区分“需要学习生成的内容”（`assistant`）和“输入上下文”（`user`/`system`） | 区分“有效 token”和“填充 token（padding）” |
| **取值含义** | `1` 表示该位置的预测错误需计入损失，`0` 表示忽略 | `1` 表示该 token 有效（参与注意力计算），`0` 表示是 padding（被忽略） |


### 举个例子
假设一段对话的 token 序列为：
```
<|im_start|>user\n你好<|im_end|><|im_start|>assistant\n你好！<|im_end|><pad><pad>
```
- **`attention_mask`**：所有非 padding 部分（`<|im_start|>user` 到 `<|im_end|>`）都为 `1`，padding 部分（`<pad>`）为 `0`。目的是让模型只关注有效对话内容，忽略填充。
- **`loss_mask`**：仅 `<|im_start|>assistant` 和 `<|im_end|>` 之间的 `你好！` 部分为 `1`，其他部分（`user` 的内容、标记符、padding）都为 `0`。目的是只让模型为生成“你好！”的错误负责，引导它学习如何回答用户问题。


综上，`loss_mask` 聚焦于“学什么”（只学生成回答），`attention_mask` 聚焦于“关注什么”（只关注有效内容），二者配合实现 SFT 任务的训练目标。
